### PR TITLE
BUG: negative spacing is not supported, but check did not enforce that

### PR DESCRIPTION
**Pull request by dzenanz**
Related to #2911 and #2917.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

File: Modules/Core/Common/include/itkImageBase.hxx [hunk 1]
```diff
@@ -72,14 +72,13 @@ ImageBase<VImageDimension>::SetSpacing(const SpacingType & spacing)
 {
   for (unsigned int i = 0; i < VImageDimension; ++i)
   {
-    if (this->m_Spacing[i] < 0.0)
+    if (spacing[i] < 0.0)
     {
 #if !defined(ITK_LEGACY_REMOVE)
-      itkWarningMacro("Negative spacing is not supported and may result in undefined behavior. Spacing is "
-                      << this->m_Spacing);
+      itkWarningMacro("Negative spacing is not supported and may result in undefined behavior. Spacing is " << spacing);
```
Comment (blowekamp) [inline Modules/Core/Common/include/itkImageBase.hxx:78]: This message should say something like:

"Negative spacing is not supported and may result in undefined behavior. Proceeding to set spacing to " <<

File: Modules/Core/Common/include/itkImageBase.hxx [hunk 2]
```diff
@@ -72,14 +72,13 @@ ImageBase<VImageDimension>::SetSpacing(const SpacingType & spacing)
...
       break;
 #else
-      itkExceptionMacro("Negative spacing is not allowed: Spacing is " << this->m_Spacing);
+      itkExceptionMacro("Negative spacing is not allowed: Spacing is " << spacing);
```
Comment (blowekamp) [inline Modules/Core/Common/include/itkImageBase.hxx:81]: This message should say something like:

"Negative spacing is not supported and may result in undefined behavior. Refusing to change spacing to " << spacing << " from " << this->m_Spacing

**Review by blowekamp — COMMENTED** [review]
Thanks for fixing the logic. The error messages need a update to be accurate and clear.

**Comment by jhlegarreta** [comment]
@dzenanz thanks for doing this.

After reading Matt's and Brad's comments in #2917, and the need to at least read negative spacing images by third party applications, maybe commit d447f0452bb5ea92a555e630d05b57da535bd3a9 was thought to be a halfway compromise before the strict enforcement, as it is guarded by the `FUTURE_LEGACY_REMOVE` macro (but I still find the message inaccurate/contradicting the fact that setting the zero/negative spacing does actually take place).

For context, maybe some related discussion about zero/negative spacings are these (?):
https://itk.org/pipermail/community/2016-November/012275.html (newest reply from François)
https://discourse.itk.org/t/nifti-file-orientation-change-between-2019-08-26-and-itk-v5-1rc01/2594/12?
https://public.kitware.com/pipermail/insight-users/2016-October/054037.html

But reading negative (at least negative) spacing images (but immediately setting to an "equivalent" positive value) is still allowed, as Brad points in https://github.com/InsightSoftwareConsortium/ITK/issues/2917#issuecomment-984794713.

Thus, now I am not sure whether I unnecessarily opened a front with my PR and thoughts. Sorry, not the most lucid days.

File: Modules/Core/Common/include/itkImageBase.hxx (hunk 2 above)
Comment (dzenanz) [reply Modules/Core/Common/include/itkImageBase.hxx]: Done in force-push.

**Comment by blowekamp** [comment]
> Thus, now I am not sure whether I unnecessarily opened a front with my PR and thoughts. Sorry, not the most lucid days.

Jon, I am glad your efforts have gotten this issue straightened out. I recall encountering it messing around with SimpleITK and setting the spacing and getting a confusing error message.

**Review by jhlegarreta — COMMENTED** [review]
The message and the check make more sense to me :-).

I guess the block should be copied and adapted to check and warn about the same situation for a zero-valued spacing, isn't it? This would mean removing (or moving to `SetSpacing` and adapting) this check:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImageBase.hxx#L163

At that point, `this->m_Spacing[i] == 0` should never evaluate to `true`.

And I guess a separate PR or another commit in this PR could do the same for zero-valued directions?
i.e. add a check here:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImageBase.hxx#L131
As well as moving this other direction-related determinant check (also checking the method parameter instead of `m_Direction`) to `SetDirection` as well ?
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImageBase.hxx#L170

**Comment by dzenanz** [comment]
@jhlegarreta I have "Allow edits and access to secrets by maintainers" enabled by default, and it is active for this PR too. Feel free to add an additional commits, or amend this one. Also, you can make a separate PR, possibly later. As the changes in this PR are minimal, I am happy to abandon it in favor of a more comprehensive one.

**Comment by jhlegarreta** [comment]
So the legacy remove build is doing its job :ok_hand:  and we have the first issue uncovered by it:
https://open.cdash.org/test/554773719

The exception seems to be thrown in here:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Registration/Metricsv4/test/itkImageToImageMetricv4Test.cxx#L597
So it looks likes some registration process does

**Comment by jhlegarreta** [comment]
> @jhlegarreta I have "Allow edits and access to secrets by maintainers" enabled by default, and it is active for this PR too. Feel free to add an additional commits, or amend this one. Also, you can make a separate PR, possibly later. As the changes in this PR are minimal, I am happy to abandon it in favor of a more comprehensive one.

Done in 8df1ea4.

**Comment by dzenanz** [comment]
I guess we should change the metric exception testing to trigger it by something other than negative spacing.

**Comment by jhlegarreta** [comment]
> Not sure about the positive spacing values that would still raise the metric exception, but there are a few other displacement field image properties that might raise an exception:

> Still expect the legacy remove build failure.

Here it is #2923. Tested locally. Looks to be good enough to raise the exception.

**Comment by dzenanz** [comment]
This now passes tests. It is ready for reviewing and merging.

**Review by blowekamp — APPROVED** [review]
Nice!

**Pull request merged**